protocol/clproto_cpp/src/clproto.cpp

```diff
@@ -0,0 +1,732 @@
+#include "clproto.h"
+
+#include <state_representation/State.hpp>
+#include <state_representation/space/SpatialState.hpp>
+#include <state_representation/space/cartesian/CartesianState.hpp>
+#include <state_representation/space/cartesian/CartesianPose.hpp>
+#include <state_representation/space/cartesian/CartesianTwist.hpp>
+#include <state_representation/space/cartesian/CartesianWrench.hpp>
+#include <state_representation/robot/Jacobian.hpp>
+#include <state_representation/robot/JointState.hpp>
+#include <state_representation/robot/JointPositions.hpp>
+#include <state_representation/robot/JointVelocities.hpp>
+#include <state_representation/robot/JointTorques.hpp>
+#include <state_representation/geometry/Shape.hpp>
+#include <state_representation/geometry/Ellipsoid.hpp>
+
+#include "state_representation/state_message.pb.h"
+#include "state_representation/state.pb.h"
+#include "state_representation/space/spatial_state.pb.h"
+#include "state_representation/space/cartesian/cartesian_state.pb.h"
+#include "state_representation/space/joint/jacobian.pb.h"
+#include "state_representation/space/joint/joint_state.pb.h"
+#include "state_representation/geometry/shape.pb.h"
+#include "state_representation/geometry/ellipsoid.pb.h"
+
+using namespace state_representation;
+
+namespace clproto {
+
+DecodingException::DecodingException(const std::string& msg) : std::runtime_error(msg) {}
+NotImplementedException::NotImplementedException(const std::string& msg) : std::runtime_error(msg) {}
+
+bool is_valid(const std::string& msg) {
+  return check_message_type(msg) != MessageType::UNKNOWN_MESSAGE;
+}
+
+MessageType check_message_type(const std::string& msg) {
+  if (proto::StateMessage message; message.ParseFromString(msg)) {
+    return static_cast<MessageType>(message.message_type_case());
+  }
+
+  /* In theory, sending "raw" messages (message types without the
+   * StateMessage wrapper) could also be supported, though it would
+   * require manually checking each case as in the following snippet:
+
+  if (proto::State message; message.ParseFromString(msg)) {
+    return MessageType::STATE_MESSAGE;
+  } else if (proto::SpatialState message; message.ParseFromString(msg)) {
+    return MessageType::SPATIAL_STATE_MESSAGE;
+  }
+
+   * Because the intention is to encode / decode messages using
+   * this library, and all encoded messages use the StateMessage
+   * wrapper, raw messages are treated as UNKNOWN at this time.
+   */
+
+  return MessageType::UNKNOWN_MESSAGE;
+}
+
+/**
+ * @brief Local encoding helper for StateType
+ * @param type The state representation StateType
+ * @return The equivalent proto StateType
+ */
+static proto::StateType encoder(const StateType& type) {
```

Comment:
If I am not mistaken, this `StateType` is a `class enum`. Couldn't we simply encode it as the enum `int` value of each field?

Reply:
Yes, to be honest I'm not that happy with the StateType enum to begin with. There are two points:

- Whey encode the StateType as an enum message and not an int field:
My intention was that the message is descriptive enough to use without installing control libraries (for later front/backend use). For this, a StateType that is just a number is not helpful, since you don't know what each number means - a named enumeration will provide the type names for the user.

- Why manually assign cases in a switch statement and not just cast from one enum to the other
I could also do a static_cast between enums here, though this requires that the underlying enum index values are consistent between the proto and the control library. It's a little less brittle to assign them explicitly like this.

But, on a larger point, the StateType is a bit messy. It is not granular in the case of frequently used messages (e.g. CARTESIANSTATE for all variants), despite being granular in other cases (PARAMETERS), and contains namespace mismatches (JACOBIANMATRIX for Jacobian, and inconsistent underscores between types).

If clproto constitutes a large enough feature to make a 4.0 release, and would already prompt a refactor of message handling in modulo / downstream users, I would like to refactor this enum in state_representation to be basically equivalent to the StateMessage enum in level of description.

Reply:
4.0 release is gonna be biiiig :rocket:

Reply:
yes that makes sense